File: PropertyTestsFs/PropertyTests.fs
module PropertyTests

open FsCheck
open FsCheck.NUnit
open Code
open System.Linq

let myConfig = Config.QuickThrowOnFailure

let unitPrice = PotterCalculator.UnitPrice

let checkPriceOfBooks titles expectedPrice =
    let actualPrice = PotterCalculator.CalculatePrice titles
    actualPrice = expectedPrice

let checkPriceOfBooksSpec gen expectedPriceFun =
    let arb = Arb.fromGen gen
    let property = Prop.forAll arb (fun titles -> checkPriceOfBooks titles (expectedPriceFun titles))
    Check.One(myConfig, property)

let checkPriceOfBooksSpecWithFixedPrice gen expectedPrice =
    checkPriceOfBooksSpec gen (fun _ -> expectedPrice)

let titlesAreDifferent titles =
    Seq.length(Seq.distinct titles) = Seq.length titles

let combinationsOfTwoDifferentTitles all =
    seq {
        for title1 in all do
        for title2 in all do
        let titles = [|title1; title2|]
        if titlesAreDifferent(titles) then yield titles
    }

let combinationsOfThreeDifferentTitles all =
    seq {
        for title1 in all do
        for title2 in all do
        for title3 in all do
        let titles = [|title1; title2; title3|]
        if titlesAreDifferent(titles) then yield titles
    }

let combinationsOfFourDifferentTitles all =
    seq {
        for title1 in all do
        for title2 in all do
        for title3 in all do
        for title4 in all do
        let titles = [|title1; title2; title3; title4|]
        if titlesAreDifferent(titles) then yield titles
    }

let combinationsOfFiveDifferentTitles all =
    seq {
        for title1 in all do
        for title2 in all do
        for title3 in all do
        for title4 in all do
        for title5 in all do
        let titles = [|title1; title2; title3; title4; title5|]
        if titlesAreDifferent(titles) then yield titles
    }

let genOneTitle = Gen.elements(HarryPotterBooks.Titles)
let genOneTitleArray = Gen.map (fun title -> Seq.singleton title |> Seq.toArray) genOneTitle
let genTwoDifferentTitles = Gen.elements(combinationsOfTwoDifferentTitles(HarryPotterBooks.Titles))
let genThreeDifferentTitles = Gen.elements(combinationsOfThreeDifferentTitles(HarryPotterBooks.Titles))
let genFourDifferentTitles = Gen.elements(combinationsOfFourDifferentTitles(HarryPotterBooks.Titles))
let genFiveDifferentTitles = Gen.elements(combinationsOfFiveDifferentTitles(HarryPotterBooks.Titles))

let genMultipleTitlesTheSame =
    gen {
        let! title = genOneTitle
        let! n = Gen.choose(0, 20)
        return Enumerable.Repeat(title, n) |> Seq.toArray
    }

let genFullyOverlappingFourDifferentTitlesPlusTwoDifferentTitles =
    gen {
        let! four = genFourDifferentTitles
        let! two = Gen.elements(combinationsOfTwoDifferentTitles(four))
        return Array.append four two
    }

[<Property>]
let ``one book``() = 
    let expectedPrice = unitPrice
    checkPriceOfBooksSpecWithFixedPrice genOneTitleArray expectedPrice

[<Property>]
let ``multiple books the same``() = 
    let expectedPriceFun = fun titles -> (Seq.length titles |> decimal) * unitPrice
    checkPriceOfBooksSpec genMultipleTitlesTheSame expectedPriceFun

[<Property>]
let ``two books different``() = 
    let expectedPrice = 2m * unitPrice * 0.95m
    checkPriceOfBooksSpecWithFixedPrice genTwoDifferentTitles expectedPrice

[<Property>]
let ``three books different``() = 
    let expectedPrice = 3m * unitPrice * 0.90m
    checkPriceOfBooksSpecWithFixedPrice genThreeDifferentTitles expectedPrice

[<Property>]
let ``four books different``() = 
    let expectedPrice = 4m * unitPrice * 0.80m
    checkPriceOfBooksSpecWithFixedPrice genFourDifferentTitles expectedPrice

[<Property>]
let ``five books different``() = 
    let expectedPrice = 5m * unitPrice * 0.75m
    checkPriceOfBooksSpecWithFixedPrice genFiveDifferentTitles expectedPrice

[<Property>]
let ``fully overlapping four different books plus two different books``() =
    let gen = genFullyOverlappingFourDifferentTitlesPlusTwoDifferentTitles
    let expectedPrice = (4m * unitPrice * 0.80m) + (2m * unitPrice * 0.95m)
    checkPriceOfBooksSpecWithFixedPrice gen expectedPrice

File: PropertyTestsCs/PropertyTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Code;
using FsCheck;
using FsCheck.Fluent;

namespace PropertyTests
{
    internal class PropertyTests
    {
        private const decimal UnitPrice = PotterCalculator.UnitPrice;

        private static bool CheckPriceOfBooks(string[] titles, decimal expectedPrice)
        {
            var actualPrice = PotterCalculator.CalculatePrice(titles);
            return actualPrice == expectedPrice;
        }

        private static Func<T, IEnumerable<T>> DontShrink<T>()
        {
            return _ => Enumerable.Empty<T>();
        }

        private static void CheckPriceOfBooksSpec(Gen<string[]> gen, Func<string[], decimal> expectedPriceFunc)
        {
            Spec
                .For(gen, titles => CheckPriceOfBooks(titles, expectedPriceFunc(titles)))
                .Shrink(DontShrink<string[]>())
                .QuickCheckThrowOnFailure();
        }

        private static void CheckPriceOfBooksSpecWithFixedPrice(Gen<string[]> gen, decimal expectedPrice)
        {
            CheckPriceOfBooksSpec(gen, _ => expectedPrice);
        }

        private static bool TitlesAreDifferent(ICollection<string> titles)
        {
            return titles.Distinct().Count() == titles.Count;
        }

        private static IEnumerable<string[]> CombinationsOfTwoDifferentTitles(string[] allTitles)
        {
            return from title1 in allTitles
                   from title2 in allTitles
                   let titles = new[] { title1, title2 }
                   where TitlesAreDifferent(titles)
                   select titles;
        }

        private static IEnumerable<string[]> CombinationsOfThreeDifferentTitles(string[] allTitles)
        {
            return from title1 in allTitles
                   from title2 in allTitles
                   from title3 in allTitles
                   let titles = new[] { title1, title2, title3 }
                   where TitlesAreDifferent(titles)
                   select titles;
        }

        private static IEnumerable<string[]> CombinationsOfFourDifferentTitles(string[] allTitles)
        {
            return from title1 in allTitles
                   from title2 in allTitles
                   from title3 in allTitles
                   from title4 in allTitles
                   let titles = new[] { title1, title2, title3, title4 }
                   where TitlesAreDifferent(titles)
                   select titles;
        }

        private static IEnumerable<string[]> CombinationsOfFiveDifferentTitles(string[] allTitles)
        {
            return from title1 in allTitles
                   from title2 in allTitles
                   from title3 in allTitles
                   from title4 in allTitles
                   from title5 in allTitles
                   let titles = new[] { title1, title2, title3, title4, title5 }
                   where TitlesAreDifferent(titles)
                   select titles;
        }

        private static readonly Gen<string> GenOneTitle = Any.ValueIn(HarryPotterBooks.Titles);
        private static readonly Gen<string[]> GenOneTitleArray = GenOneTitle.Select(s => new[]{s});
        private static readonly Gen<string[]> GenTwoDifferentTitles = Any.ValueIn(CombinationsOfTwoDifferentTitles(HarryPotterBooks.Titles));
        private static readonly Gen<string[]> GenThreeDifferentTitles = Any.ValueIn(CombinationsOfThreeDifferentTitles(HarryPotterBooks.Titles));
        private static readonly Gen<string[]> GenFourDifferentTitles = Any.ValueIn(CombinationsOfFourDifferentTitles(HarryPotterBooks.Titles));
        private static readonly Gen<string[]> GenFiveDifferentTitles = Any.ValueIn(CombinationsOfFiveDifferentTitles(HarryPotterBooks.Titles));

        private static readonly Gen<string[]> GenMultipleTitlesTheSame =
            from title in GenOneTitle
            from n in Any.IntBetween(0, 20)
            select Enumerable.Repeat(title, n).ToArray();

        private static readonly Gen<string[]> GenFullyOverlappingFourDifferentTitlesPlusTwoDifferentTitles =
            from four in GenFourDifferentTitles
            from two in Any.ValueIn(CombinationsOfTwoDifferentTitles(four))
            select four.Concat(two).ToArray();

        [FsCheck.NUnit.Property]
        public void OneBook()
        {
            const decimal expectedPrice = UnitPrice;
            CheckPriceOfBooksSpecWithFixedPrice(GenOneTitleArray, expectedPrice);
        }

        [FsCheck.NUnit.Property]
        public void MultipleBooksTheSame()
        {
            CheckPriceOfBooksSpec(GenMultipleTitlesTheSame, titles => titles.Length * UnitPrice);
        }

        [FsCheck.NUnit.Property]
        public void TwoDifferentBooks()
        {
            const decimal expectedPrice = 2 * UnitPrice * 0.95m;
            CheckPriceOfBooksSpecWithFixedPrice(GenTwoDifferentTitles, expectedPrice);
        }

        [FsCheck.NUnit.Property]
        public void ThreeDifferentBooks()
        {
            const decimal expectedPrice = 3 * UnitPrice * 0.90m;
            CheckPriceOfBooksSpecWithFixedPrice(GenThreeDifferentTitles, expectedPrice);
        }

        [FsCheck.NUnit.Property]
        public void FourDifferentBooks()
        {
            const decimal expectedPrice = 4 * UnitPrice * 0.80m;
            CheckPriceOfBooksSpecWithFixedPrice(GenFourDifferentTitles, expectedPrice);
        }

        [FsCheck.NUnit.Property]
        public void FiveDifferentBooks()
        {
            const decimal expectedPrice = 5 * UnitPrice * 0.75m;
            CheckPriceOfBooksSpecWithFixedPrice(GenFiveDifferentTitles, expectedPrice);
        }

        [FsCheck.NUnit.Property]
        public void FullyOverlappingFourDifferentBooksPlusTwoDifferentBooks()
        {
            var gen = GenFullyOverlappingFourDifferentTitlesPlusTwoDifferentTitles;
            const decimal expectedPrice = (4 * UnitPrice * 0.80m) + (2 * UnitPrice * 0.95m);
            CheckPriceOfBooksSpecWithFixedPrice(gen, expectedPrice);
        }
    }
}
